streams/jvm/src/main/scala/zio/stream/compression/Gunzipper.scala

```diff
@@ -0,0 +1,184 @@
+package zio.stream.compression
+
+import java.util.zip.{ CRC32, DataFormatException, Inflater }
+import java.{ util => ju }
+
+import scala.annotation.tailrec
+
+import zio._
+
+private[compression] class Gunzipper private (var state: Gunzipper.State) {
+  def onChunk(c: Chunk[Byte]): ZIO[Any, Throwable, Chunk[Byte]] =
+    state.feed(c).map {
+      case (newState, output) =>
+        state = newState
+        output
+    }
+
+  def close: UIO[Unit] = state.close
+}
+
+private[compression] object Gunzipper {
+
+  private val fixedHeaderLength = 10
+
+  sealed trait State {
```

Comment:
I love how you organized the steps here. I'm not knowledgeable about the GZIP encoding, so I won't review everything here, but it looks great from the maintenance angle.

Hopefully ZIO Codec will make such parsers easier to write 💪🏻